### PR TITLE
[feat]: add experimental WorkflowUI telemetry hooks

### PR DESCRIPTION
**Pull request by jamieQ**
### Motivation

we'd like to be able to build out some observation tooling that is aware of certain events that occur within WorkflowUI. specifically, view lifecycle data is of particular interest. currently the options for tying into UIViewController lifecycle events are either:

1. introduce new subclasses in client code that override the methods of interest and provide observation hooks there.
2. swizzle the methods of interest to allow client observation

option `1.` may place a large burden on consumers that have many existing subclasses (of, say, ScreenViewController), and only works for the `open` types defined in `WorkflowUI`, which excludes `WorkflowHostingController` & `DescribedViewController`. option `2.` may require less code to implement, but adds indirection (swizzling is often not particularly discoverable), runtime overhead, and may not work for Swift-only methods.

note: this is all currently implemented behind 'experimental' SPI declarations. the intent would be to treat this as sort of an 'alpha' version of this functionality, and to set expectations that it may have breaking changes going forward, without a major version increment in the library.

### Implementation

to achieve this goal, we've implemented an approach that does the following:

- unifies all `UIViewController` subtypes defined in `WorkflowUI` under a new parent class `WorkflowUIViewController`
- instruments some view controller lifecycle events (viewDidAppear, etc) within the shared base class
- exposes a static value that can be used to register a 'global' observer of these events
- adds some new protocols and event data types to enable observation of WorkflowUI lifecycle events

### Usage Example

the envisioned consumer is something like a shared telemetry system that wishes to observe certain WorkflowUI events and directly use or attach metadata to them. for example, imagine a consumer wished to log an event whenever a ViewController from WorkflowUI appeared on screen. with the proposed changes, this could be accomplished as follows:

```swift
// implement an observer
struct AppearanceLogger: WorkflowUIObserver {
    var loggingSystem: LoggingSystem

    // handle all emitted observation events
    func observeEvent<E: WorkflowUIEvent>(_ event: E) {
        // perform specific observation logic
        if let appearanceEvent = event as? ViewDidAppearEvent {
            loggingSystem.logAppearanceEvent(appearanceEvent.viewController)
        }
    }
}

// install the global observer
WorkflowUIObservation.sharedUIObserver = AppearanceLogger()
```

## Checklist

- [x] Unit Tests
- [x] UI Tests
- [x] Snapshot Tests (iOS only)
- [x] I have made corresponding changes to the documentation

### CODE REVIEW

File: WorkflowUI/Sources/Observation/WorkflowUIEvents.swift [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#if canImport(UIKit)
+import Foundation
+import UIKit
+
+public protocol WorkflowUIEvent {
```
Comment (jamieQ) [inline WorkflowUI/Sources/Observation/WorkflowUIEvents.swift:21]: looking for feedback on the general structuring of the event types and related protocols

File: WorkflowUI/Sources/Hosting/WorkflowHostingController.swift [hunk 2]
```diff
@@ -150,4 +155,23 @@ public final class WorkflowHostingController<ScreenType, Output>: UIViewControll
     }
 }
 
+// MARK: Observation Events
+
+public protocol WorkflowHostingControllerEvent: WorkflowUIEvent {
```
Comment (jamieQ) [inline WorkflowUI/Sources/Hosting/WorkflowHostingController.swift:160]: i'm not sure how useful this protocol is... since you have to know all the generic types up front, there isn't really a good way to do something like: 'given this `WorkflowUIEvent`, conditionally do something if it was emitted from a WorkflowHostingController'. you basically can't know that the event came from a hosting controller unless we were to add either some shared metadata to the base event protocol that specifies that, or make a marker protocol that ignores the associated types (or something else?).

File: WorkflowUI/Tests/WorkflowUIObservationTestCase.swift [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import Combine
+import Workflow
+@_spi(WorkflowUIGlobalObservation) import WorkflowUI
+import XCTest
+
+open class WorkflowUIObservationTestCase: XCTestCase {
```
Comment (jamieQ) [inline WorkflowUI/Tests/WorkflowUIObservationTestCase.swift:22]: there's some scratch work in here, but some ideas for how one could tie into this system and forward events through Combine

File: WorkflowUI/Sources/Observation/WorkflowUIEvents.swift [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    var viewController: UIViewController { get }
+}
+
+// MARK: ViewController Lifecycle Events
+
+public protocol ViewControllerLifeCycleEvent: WorkflowUIEvent {}
```
Comment (amorde) [inline WorkflowUI/Sources/Observation/WorkflowUIEvents.swift:27]: I see the intent but since `WorkflowUIEvent` always has a `UIViewController`, does this add any value?

File: WorkflowUI/Sources/Observation/WorkflowUIViewController.swift [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+        sendObservationEvent(ViewDidAppearEvent(
+            viewController: self,
+            animated: animated,
+            isFirstAppearance: isFirstAppearance
+        ))
+    }
+
+    override open func viewWillLayoutSubviews() {
+        // no need to call super since it does nothing
```
Comment (amorde) [inline WorkflowUI/Sources/Observation/WorkflowUIViewController.swift:61]: technically correct, though it does feel a bit wrong..

**Review by amorde — APPROVED** [review]

File: WorkflowUI/Sources/Hosting/WorkflowHostingController.swift (hunk 2 above)
Comment (amorde) [reply WorkflowUI/Sources/Hosting/WorkflowHostingController.swift:160]: Agree, this doesn't seem to add value other than enforcing some naming conventions for the generic types

File: WorkflowUI/Sources/Observation/WorkflowUIEvents.swift (hunk 1 above)
Comment (amorde) [reply WorkflowUI/Sources/Observation/WorkflowUIEvents.swift:22]: Structure looks good to me

File: WorkflowUI/Sources/Observation/WorkflowUIViewController.swift [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#if canImport(UIKit)
+import Foundation
+import UIKit
+
+/// Ancestor type from which all ViewControllers in WorkflowUI inherit.
+open class WorkflowUIViewController: UIViewController {
```
Comment (jamieQ) [inline WorkflowUI/Sources/Observation/WorkflowUIViewController.swift:22]: cc @n8chur i know you've got a feature branch that this might conflict with. any issues with adding this new common supertype you can think of?

**Review by mjohnson12 — APPROVED** [review]
This looks good to me and flows well with the rest of the WorkflowUI code.

**Review by Kiesco08 — APPROVED** [review]
Approach lgtm too, seems easy to use

File: WorkflowUI/Sources/Hosting/WorkflowHostingController.swift [hunk 7]
```diff
@@ -99,7 +99,7 @@ public final class WorkflowHostingController<ScreenType, Output>: UIViewControll
     }
 
     override public func viewDidLayoutSubviews() {
-        super.viewDidLayoutSubviews()
+        defer { super.viewDidLayoutSubviews() }
```
Comment (watt) [inline WorkflowUI/Sources/Hosting/WorkflowHostingController.swift:102]: Is this change needed for your hooks to work? We were considering moving the frame assignment to `viewWillLayoutSubviews` as part of the `ViewEnvironmentUI` feature, but wanted to isolate that change due to potential for side effects in code that was depending on this order of events.

**Review by square-tomb — APPROVED** [review]
Looks good to me!

WorkflowSwiftUI contains a type `WorkflowHostingViewController` that should probably now subclass `WorkflowUIViewController`, too.

Other approaches to SwiftUI support that we're experimenting with involve [a view controller type that sublclasses `UIHostingController`](https://github.com/square/workflow-swift/blob/93c3b5d281bff9de2e40390c1192a25dfe8a7a45/WorkflowSwiftUI/Sources/SwiftUIScreen.swift#L83) and so cannot subclass `WorkflowUIViewController`.

However, I think any view controller type like that can be modified to contain, rather than subclass, `UIHostingController`. The only downside AFAIK would be a deeper VC hierarchy.

File: WorkflowUI/Sources/Hosting/WorkflowHostingController.swift (hunk 2 above)
Comment (jamieQ) [reply WorkflowUI/Sources/Hosting/WorkflowHostingController.swift]: remove this for now, and may revisit in the future

File: WorkflowUI/Sources/Hosting/WorkflowHostingController.swift (hunk 7 above)
Comment (jamieQ) [reply WorkflowUI/Sources/Hosting/WorkflowHostingController.swift:102]: no i don't think this change is necessary. the previous implementation of `super.viewDidLayoutSubviews()` did nothing (according to the docs), and the new one emits an event. it seems like the event emission should occur after any logic in this method, but i don't think the ordering relative to the frame assignment will matter to any of the envisioned consumers.

if we're potentially going to move the frame assignment out of this method, then perhaps we should leave this proposed change since it will preserve the relative ordering with the new observation events?

File: WorkflowUI/Tests/WorkflowUIViewControllerTests.swift [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+    var value: Int = 0
+
+    func viewControllerDescription(environment: ViewEnvironment) -> ViewControllerDescription {
+        .testing
+    }
+}
+
+private extension ViewControllerDescription {
+    static var testing: Self {
+        .init(build: { UIViewController() }, update: { _ in })
```
Comment (n8chur) [inline WorkflowUI/Tests/WorkflowUIViewControllerTests.swift:236]: I think this would be the only conflict when compared to the ViewEnvironmentUI work—`ViewControllerDescription` requires an `environment` parameter in that feature branch. It should be easy to reconcile depending on who lands their work first!

**Review by n8chur — APPROVED** [review]
I'm a bit nervous about the reliance on a `UIViewController` superclass when `ViewControllerDescription` does not enforce this, but it seems like it may be worth the alternative given our goals!

`ViewControllerDescriptions` not describing a `ScreenViewController` subclass does seem like a potential footgun here (what @square-tomb pointed out in a previous comment).

Are we planning on adding documentation that describes in detail how to adhere to these event contract rules once it's out of alpha? Could Swift macros help make adding adherence easier in the future without needing to wrap one VC in another? It'd be nice to avoid needing to compose a nested VC to get this behavior.

**Comment by kyleve** [comment]
> ViewControllerDescriptions not describing a ScreenViewController subclass does seem like a potential footgun here

Agreed. I don't think this is something we can/should enforce here, IMO.

File: WorkflowUI/Sources/Hosting/WorkflowHostingController.swift [hunk 9]
```diff
@@ -22,7 +22,7 @@ import UIKit
 import Workflow
 
 /// Drives view controllers from a root Workflow.
-public final class WorkflowHostingController<ScreenType, Output>: UIViewController where ScreenType: Screen {
+public final class WorkflowHostingController<ScreenType, Output>: WorkflowUIViewController where ScreenType: Screen {
```
Comment (jamieQ) [inline WorkflowUI/Sources/Hosting/WorkflowHostingController.swift:25]: @square-tomb responding here for threading purposes:

> WorkflowSwiftUI contains a type `WorkflowHostingViewController` that should probably now subclass `WorkflowUIViewController`, too.

good callouts, but given the current motivations for this interface, i think it's okay to punt on the story with SwiftUI and ignore this for the time being.

Comment (jamieQ) [inline WorkflowUI/Sources/Hosting/WorkflowHostingController.swift:25]: @kyleve responding here for threading purposes:

> >ViewControllerDescriptions not describing a ScreenViewController subclass does seem like a potential footgun here
> 
> Agreed. I don't think this is something we can/should enforce here, IMO.

could you clarify – you think this is a risk, but not one we should change the contract of ViewControllerDescription to deal with?

Comment (jamieQ) [reply WorkflowUI/Sources/Hosting/WorkflowHostingController.swift:25]: > I wonder, do we need events from`ScreenViewController` at all? Maybe the events from `DescribedViewController` are sufficient if that VC usually maps 1:1 to a screen.

being able to recover the Screen type from ScreenViewController may be of use

> If not, and `DescribedViewController` is often used to display a succession of different screens, we would have to add some events [here](https://github.com/square/workflow-swift/pull/221/files#diff-20be64c4e0d93842089c1503234a2662af08e4a919c9f946e737e94015f5e47bR45-R47). Or we could wrap `currentViewController` in a container `WorkflowUIViewController` if we don't mind deepening the VC hierarchy.

adding some additional events that are more specific to the logic in some of the VC subtypes may be something we want to do eventually, but for now i thought we'd just start with VC lifecycle information and see how far that gets us.

[the remark this answers is not in the document]
Comment (n8chur) [reply WorkflowUI/Sources/Hosting/WorkflowHostingController.swift:25]: > i thought macros had to be additive, so i'm not sure we'd be able to change any existing code in client-supplied VCs.

I dove a little deeper into Swift Macros after my initial comment and I think you're right, it looks like it may not be useful here.
